File: infrastructure/src/main/java/com/study/platform/global/idempotency/IdempotencyObjectStorageAdapter.java
package com.study.platform.global.idempotency;

import lombok.RequiredArgsConstructor;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Component;

import java.time.Duration;
import java.util.concurrent.TimeUnit;

@Component
@RequiredArgsConstructor
public class IdempotencyObjectStorageAdapter implements IdempotencyObjectStoragePort {

    private final RedisTemplate<String, Object> redisTemplate;

    @Override
    public Object get(String key) {
        return redisTemplate.opsForValue().get(key);
    }

    @Override
    public boolean setIfAbsent(String key, String marker, Duration ttl) {
        return Boolean.TRUE.equals(
                redisTemplate.opsForValue().setIfAbsent(key, marker, ttl.toSeconds(), TimeUnit.SECONDS));
    }

    @Override
    public void set(String key, Object value, Duration ttl) {
        redisTemplate.opsForValue().set(key, value, ttl);
    }

    @Override
    public void delete(String key) {
        redisTemplate.delete(key);
    }
}

File: infrastructure/src/main/kotlin/com/study/platform/global/idempotency/IdempotencyObjectStorageAdapter.kt
package com.study.platform.global.idempotency

import org.springframework.data.redis.core.RedisTemplate
import org.springframework.stereotype.Component
import java.time.Duration
import java.util.concurrent.TimeUnit

@Component
class IdempotencyObjectStorageAdapter(
    private val redisTemplate: RedisTemplate<String, Any>
) : IdempotencyObjectStoragePort {

    override fun get(key: String): Any? =
        redisTemplate.opsForValue().get(key)

    override fun setIfAbsent(key: String, marker: String, ttl: Duration): Boolean =
        redisTemplate.opsForValue().setIfAbsent(key, marker, ttl.toSeconds(), TimeUnit.SECONDS) == true

    override fun set(key: String, value: Any, ttl: Duration) {
        redisTemplate.opsForValue().set(key, value, ttl)
    }

    override fun delete(key: String) {
        redisTemplate.delete(key)
    }
}
